infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/environment/convert/ComposeEnvironmentConverter.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.environment.convert;
+
+import com.google.common.collect.Maps;
+import java.util.Map;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironment;
+import org.eclipse.che.workspace.infrastructure.docker.environment.compose.ComposeEnvironment;
+import org.eclipse.che.workspace.infrastructure.docker.environment.compose.model.ComposeService;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerBuildContext;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerEnvironment;
+
+/**
+ * Converts {@link ComposeEnvironment} to {@link DockerEnvironment}.
+ *
+ * @author Alexander Garagatyi
+ * @author Alexander Andrienko
+ */
+public class ComposeEnvironmentConverter implements DockerEnvironmentConverter {
```

Comment:
Why we need that? 
Make the object you need for runtime inside InternalEnvironment with factory.

Reply:
The current approach doesn't link environments representations to a specific infra. Then, say, OpenShift infra can convert it how it prefers and doesn't have to convert Docker infra's internal representation into its own but instead converts recipe specific representation into its own. 
Personally, I like the current approach more.

Reply:
DockerEnvironment is not a common environment that can be used by any infra because it doesn't have the corresponding recipe type.
As Oleksandr said It gives an ability for implementation for infrastructure to convert environment as it wants.
OpenShift infrastructure will support DockerImageEnvironments and to implement such functionality it requires the only value of docker image but not converted DockerImageEnvironments into DockerEnvironments with containers configs and docker specific information.
Hope I described clearly why we need DockerEnvironmentConverter inside docker infra.

Reply:
I think you missed the point. I am not talking about universal object, sure OpenShift environment will have to adopt the stuffs came from Docker Environment.
What I mean is that concrete InternalEnvironment should contain object good enough for Infrastructure to prepare RuntimeContext w/o any intermediate abstractions/processing.
I.e. if ComposeEnvironment is what prepare() method expects - use it. If "current" implementations due to historical reason need DockerEnvironment - convert it inside RuntimeContext.
It will do the same job with less code and more consistency.

Reply:
Docker env accepts Dockerimage, Dockerfile and Compose internal environments. All of them have different models. To have simpler code we need some way to use them in the same way - convert into the same format. 
Both Docker and OpenShift infras need that. But both of them have their own universal format. 
I don't see that way to use all of the environments in both infras and decrease the amount of converting and simplify the code by removing some converters. 
@gazarenkov Do you have any particular example that can demonstrate how to simplify the code by removing converters? This might help to understand your idea of simplification.

Reply:
@garagatyi I perfectly understand what you did but still think you have enough abstractions to not to create one more.
Options:
- Option number 1 (what I did to not to change your code): produce DockerContainers in every Docker*InternalEnv. Pass it to OpenshiftInfra.prepare() - what is the problem? You have all you need.
- Option number 2: leave as it is for Docker*InternalEnv and perform what you did  in Converter inside Infra.prepare 
- I also have option 3 but it requires more reworking so to discuss in future, maybe.

Reply:
I think option 2 is already in this PR.
With option 1 we would still need to convert compose to DockerInternalEnvironment and then to OpenShiftInternalEnvironment. So I don't see simplification here.
Maybe the option you have in mind and requires refactoring would simplify code but as you said it requires a discussion.